pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java

```diff
@@ -0,0 +1,302 @@
+            continue;
+          }
+
+          // Replicas in ONLINE state are always allowed
+          if (state.equalsIgnoreCase(CommonConstants.Helix.StateModel.RealtimeSegmentOnlineOfflineStateModel.ONLINE)) {
+            validReplicas.add(instance);
+            continue;
+          }
+
+          // Replicas in CONSUMING state are only allowed on the last segment
+          if (state.equalsIgnoreCase(CommonConstants.Helix.StateModel.RealtimeSegmentOnlineOfflineStateModel.CONSUMING)
+              && segment.equals(validConsumingSegment)) {
+            validReplicas.add(instance);
+          }
+        }
+
+        segmentToReplicaSetQueue.add(new ImmutablePair<String, Set<String>>(segment, validReplicas));
+      }
+    }
+
+    // 4. For each segment to be used during assignment, pick a random replica, weighted by the number of segments
+    //    assigned to each replica.
+    List<ServerToSegmentSetMap> routingTables = new ArrayList<ServerToSegmentSetMap>(routingTableCount);
+    for(int i = 0; i < routingTableCount; ++i) {
+      Map<String, Set<String>> instanceToSegmentSetMap = new HashMap<String, Set<String>>();
+
+      PriorityQueue<Pair<String, Set<String>>> segmentToReplicaSetQueueCopy =
+          new PriorityQueue<Pair<String, Set<String>>>(segmentToReplicaSetQueue);
+
+      while (!segmentToReplicaSetQueueCopy.isEmpty()) {
+        Pair<String, Set<String>> segmentAndValidReplicaSet = segmentToReplicaSetQueueCopy.poll();
+        String segment = segmentAndValidReplicaSet.getKey();
+        Set<String> validReplicaSet = segmentAndValidReplicaSet.getValue();
+
+        String replica = pickWeightedRandomReplica(validReplicaSet, instanceToSegmentSetMap);
+        if (replica != null) {
+          Set<String> segmentsForInstance = instanceToSegmentSetMap.get(replica);
+
+          if (segmentsForInstance == null) {
+            segmentsForInstance = new HashSet<String>();
+            instanceToSegmentSetMap.put(replica, segmentsForInstance);
+          }
+
+          segmentsForInstance.add(segment);
+        }
+      }
+
+      routingTables.add(new ServerToSegmentSetMap(instanceToSegmentSetMap));
+    }
+
+    return routingTables;
+  }
+
+  private String pickWeightedRandomReplica(Set<String> validReplicaSet,
+      Map<String, Set<String>> instanceToSegmentSetMap) {
+    Random random = new Random();
+
+    // No replicas?
+    if (validReplicaSet.isEmpty()) {
+      return null;
+    }
+
+    // Only one valid replica?
+    if (validReplicaSet.size() == 1) {
+      return validReplicaSet.iterator().next();
+    }
+
+    // Find maximum segment count assigned to a replica
+    String[] replicas = validReplicaSet.toArray(new String[validReplicaSet.size()]);
+    int[] replicaSegmentCounts = new int[validReplicaSet.size()];
+
+    int maxSegmentCount = 0;
+    for (int i = 0; i < replicas.length; i++) {
+      String replica = replicas[i];
+      int replicaSegmentCount = instanceToSegmentSetMap.get(replica).size();
```

Comment:
wont you get an NPE here initially? since instanceToSegmentSetMap.get(replica) is null to start with?